### ISSUE TITLE
HasQueryFilters cant be used outside DbContext scope

### ISSUE THREAD

**Issue by AndersMalmgren** [aspnet/EntityFrameworkCore#18417]
For a enterprise level application you cant define an entire data model directly on the DbContext, you use `IEntityTypeConfiguration<TEntity>` to seperate configuration away from `OnModelCreating` method. 

When it comes to `HasQueryFilter` that uses expressions with dependencies to scoped members there is no such mechanics. It needs to be defined directly on the `DbContext`. This just wont cut it for a large Enterprise level application.

**Comment by smitpatel** [comment]
Your assumption is wrong. There are various different ways in which you can define Query filter outside of DbContext and still use it to change value of parameters.
Out tests here covers a large pattern which is supported.

https://github.com/aspnet/EntityFrameworkCore/blob/37de58b3addf44e83fe0fefbd5be13c39bed3357/test/EFCore.Specification.Tests/TestModels/QueryFilterFuncletizationContext.cs#L33-L60

**Comment by AndersMalmgren** [comment]
But the fields are defined on DbContext, the fields need to be defined on the custom class for true seperation. For examaple. I want todo something like this

```csharp
public class AgreementAuthorization : IEntityAuthorization
{
    private readonly string _legalEntityNumber;

    public AgreementAuthorization(IAuthScope scope) //Scoped through scoped service provider
    {
        _legalEntityNumber = scope.LegalEntityNumber; 
    }

    public void Build(ModelBuilder builder)
    {
        builder
            .Entity<Agreement>()
            .HasQueryFilter(a => _legalEntityNumber == null || a.LegalEntity.Number == _legalEntityNumber);
    }
}

public class MyDbContext : DbContext 
{
   private IEnumerable<IEntityAuthorization> _entityAuthorization;

   MyDbContext(IEnumerable<IEntityAuthorization> entityAuthorization) 
   {
      _entityAuthorization = entityAuthorization;
   }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {                                
        builder.ApplyConfigurationsFromAssembly(typeof(MyDbContext).Assembly);

        _entityAuthorization.ForEach(a => a.Build(builder));
    }
}
```

**Comment by smitpatel** [comment]
First you are defining QueryFilter for same EntityType repeatedly so that will be overwritten.

Even if I ignore that, can you describe how do you want us to figure out exact value to use at the runtime in the filter when executing query? Currently property/field/methods are defined on DbContext because we know how to inject DbContext which is executing query to get actual value when running the query.

**Comment by AndersMalmgren** [comment]
No I am not, they are injected using IoC

```csharp
service.AddTransiant<IEntityAuthorization, AgreementAuthorization>();
service.AddTransiant<IEntityAuthorization, OtherAuth>();
service.AddTransiant<IEntityAuthorization, SomeAuth>();
```

**Comment by AndersMalmgren** [comment]
Why not have hooks that trigger before queries.

**Comment by smitpatel** [comment]
> _entityAuthorization.ForEach(a => a.Build(builder));

This thing calls HasQueryFilter on `Agreement` entity multiple times. Last one wins!

**Comment by AndersMalmgren** [comment]
No it calls `HasQueryFilter `on what ever you define in `IEntityAuthorization.Build`

**Comment by AndersMalmgren** [comment]
```
HasQueryHook<TEntity>(Func<IQueryable<TEntity, IQueryable<TEntity>> queryHook);
````
Internally called when doing `DbSet<TEntity>` and filters dbset before its exposed to the user

**Comment by ajcvickers** [comment]
Notes from triage: the current design of query filters uses the `DbContext` as the anchor from which filter parameters are read. This works well in many cases, but has a couple of limitations:
* People don't get it. That is, many times we see code that attempts to hard-code the filter value into the model and then re-build the model for each context instance. This can be made to work, but with terrible perf characteristics, which makes it a pit-of-failure.
* It requires a stronger coupling of the `DbContext` type to the entity types being used that is ideal--essentially the issue reported above.

Based on this, it might be worth designing and implementing query filters where the filter parameter is obtained in a different way--such as through a defined callback. This would be easier for many to understand and would not have a strong coupling to the context.

**Comment by AndersMalmgren** [comment]
We have already abstracted the Dbcobtext away from the domain it uses a custom interface with a custom Set method so I might just write my own mechanic that hooks in our custom Set method. Only downside is that it will return a IQueryable instad of DbSet. But I can around that by making sure evybody uses AddRange and similar on the abstracted dbcobtext rather than the DbSet

**Comment by joshcomley** [comment]
@AndersMalmgren this won't apply the query filters to any `.Include(..)` you use. Just want you to be aware of the security risk of doing that; your API may expose unwanted data via expands.

**Comment by AndersMalmgren** [comment]
> @AndersMalmgren this won't apply the query filters to any `.Include(..)` you use. Just want you to be aware of the security risk of doing that; your API may expose unwanted data via expands.

Good to know, we don't use include we project our entities to DTOs. But it's a bit fragile I must admit maybe better I write something of my own. Thanks

**Comment by AndersMalmgren** [comment]
No, If any entity has a relation to another entity that is under restriction that entity also needs to have a filters. Thats ok in our case. Child and parents always have same restrictions.
